llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -0,0 +1,257 @@
+//===- sycl-post-link.cpp - SYCL post-link device code processing tool ----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This utility splits an input module into smaller ones.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/ADT/SetVector.h"
+#include "llvm/Bitcode/BitcodeWriterPass.h"
+#include "llvm/IR/IRPrintingPasses.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IRReader/IRReader.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/InitLLVM.h"
+#include "llvm/Support/SystemUtils.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+#include "llvm/Transforms/IPO.h"
+#include <memory>
+
+using namespace llvm;
+
+cl::OptionCategory ExtractCat{"sycl-post-link Options"};
```

Comment:
I think we also wanted to implement a 1 kernel - 1 module mode under a special option. Should be small update to the code.

Reply:
I wanted to implement basic functionality first so didn't add this feature. But, you are right, this is a small update. Implemented and added a test.

Reply:
Thank you!